### PR TITLE
Build payments summary / auto renewal toggle

### LINKED ISSUE #8965: UA subscriptions: Build display of latest/next/summarised payments
Display the summary of payments, and the previous/next payment details, for subscriptions as [described in the specification](https://docs.google.com/document/d/1K65n3NpZCh7EWDXKLhG31seGJ2jKbW6lqw3BwV2mlnU/edit#heading=h.q653zkl5xi23).

(This was [designed in web-squad#3502](https://github.com/canonical-web-and-design/web-squad/issues/3502).)

### PR DESCRIPTION
**Pull request by ClementChaumel**
## Done

- Added a subscription summary above the table.
- Added a collapsible menu to toggle on and off auto renewal for your monthly subs

## To do

- [ ]  Create the backend endpoint to update auto renewal status

## QA
First you need to add some monthly subscriptions to your account if you don't have some yet:
- Connect to the VPN
- Go here http://10.48.130.194:8080/staging/accounts/
- Connect with your SSO account
- Add some monthly subscriptions to your account

When you have monthly subscriptions:
- Go here https://ubuntu-com-9095.demos.haus/advantage?test_backend=true
- Login with your SSO account
- Check that the summary is correct

## Issue / Card

Fixes #8965 

## Screenshots
image

### CODE REVIEW

File: webapp/views.py [hunk 1]
```diff
@@ -451,6 +453,24 @@ def advantage_view():
         for account in accounts:
             account["contracts"] = advantage.get_account_contracts(account)
 
+            if account is not None:
```
Comment (jpmartinspt) [inline webapp/views.py:456]: Can an account be `None`? I think if one was to be the program would blow up on line `454`

File: webapp/views.py [hunk 2]
```diff
@@ -397,6 +397,8 @@ def advantage_view():
     personal_account = None
     enterprise_contracts = {}
     entitlements = {}
+    subscriptions = {}
```
Comment (jpmartinspt) [inline webapp/views.py:400]: one line this with `subscriptions = { "total_subscriptionns": 0 }`

File: webapp/views.py [hunk 3]
```diff
@@ -451,6 +453,24 @@ def advantage_view():
...
             account["contracts"] = advantage.get_account_contracts(account)
 
+            if account is not None:
+                resp = advantage.get_account_subscriptions_for_marketplace(
+                    account["id"], "canonical-ua"
+                )
+                subs = resp.get("subscriptions")
+                if subs is not None:
+                    subscriptions["current_subscription_number"] = len(subs[0]["purchasedProductListings"])
+                    subscriptions["next_payment"] = {}
```
Comment (jpmartinspt) [inline webapp/views.py:463]: if you initialise `next_payment` at the top too the first subscription does not need a special case I think.

**Comment by matthewpaulthomas** [comment]
My last day, I finally got a reason to set up the Canonical VPN. 🙃

> Add some monthly subscriptions to your account

I tried doing this the first time without entering card details, got a Stripe error, filled in the card details, got an error about a “pending purchase”, then when I tried to “resolve” it, I got stuck on a screen that forever says “Loading…” (even after reloading).

image

Meanwhile, one quick point from the screenshot:

image

The label for something should be nearer that thing than anything else. Here, the labels are clumped together and the data is clumped together. Can we have less space between labels and data, and more space between lines? The expanded plan-name cells in the table get this right, for example:

image

Thanks!

**Comment by elioqoshi** [comment]
@ClementChaumel once you address @matthewpaulthomas last comments, I could give it another UX review

**Comment by alnvdl** [comment]
Some general notes (and questions):

- A customer may have two or more types of subscriptions. At the moment, we anticipate having one that is monthly, and another that is yearly. Both may be active at the same time. How would these alternatives be presented in the payments view being introduced here?

- A subscription affects a contract by adding contract items to it (think of the subscription as "adding credits" to a contract). So the same contract may be affected by multiple subscriptions (e.g., ua-desktop-essential contract gets items coming from monthly and yearly subscriptions at the same time). Ideally, the things one has (contracts) should be displayed separately from the subscriptions. Also, how could we display the things a subscription added to a contract (its contract items)?

- When purchasing more things for a subscription, there should be a clear way to identify the period you want (monthly/yearly). This will affect the product listings that are available. How would the purchase wizard adapt to this?

File: webapp/views.py (hunk 1 above)
Comment (ClementChaumel) [reply webapp/views.py]: Good catch, I removed it

File: templates/advantage/index.html [hunk 4]
```diff
@@ -40,14 +40,34 @@ <h1>Ubuntu Advantage for&nbsp;Infrastructure</h1>
     <section class="p-strip is-shallow" style="overflow: visible;">
       <div class="row u-vertically-center">
         <div class="col-6 col-medium-3">
-          <h2>Your paid subscriptions</h2>
+          <h2>Your {% if subscriptions['total_subscriptions'] > 1 %}{{subscriptions['total_subscriptions']}} {%endif%}paid subscription{% if subscriptions['total_subscriptions'] == 1 %}{%else%}s{%endif%}</h2>
```
Comment (anthonydillon) [inline templates/advantage/index.html:43]: Tiny comment but could do with spaces between the `%` and the operator. For example: `{% endif %}`

File: templates/advantage/index.html [hunk 5]
```diff
@@ -40,14 +40,34 @@ <h1>Ubuntu Advantage for&nbsp;Infrastructure</h1>
...
         </div>
         <div class="col-6 col-medium-3 u-align--right">
           <div id="reveal-table">
-            <a href="#reveal-table" class="p-button u-no-margin--bottom" id="js-reveal-table" aria-controls="enterprise-table" aria-expanded="false">Add another subscription <i class="p-icon--contextual-menu">Open</i></a>
+            <a href="#reveal-table" class="p-button u-no-margin--bottom js-reveal-table" aria-controls="enterprise-table" aria-expanded="false">Add another subscription <i class="p-icon--contextual-menu">Open</i></a>
           </div>
         </div>
       </div>
+      {%if subscriptions['has_monthly'] %}
```
Comment (anthonydillon) [inline templates/advantage/index.html:51]: Same here, just needs a space

**Comment by SirSamTumless** [comment]
@wgx Can you confirm this is still needed?

**Review by anthonydillon — APPROVED** [review]
LGTM with a few niks.

I agree we could put the same content in the button and remove the dropdown and make it just a sentence. Also, from a UX point this could be misleading as you could have multiple payments in one month but we only display the last date.

**Comment by wgx** [comment]
@SirSamTumless 

A chevron that exposes an extra panel of information.. it's a bit weird but not terrible - people will understand this I'm sure. 

This falls into the category of "lets fix later" as part of the redesign for this page. 

I vote leave it for now

File: templates/advantage/index.html (hunk 4 above)
Comment (ClementChaumel) [reply templates/advantage/index.html]: Done! :+1:

**Comment by anasereijo** [comment]
@ClementChaumel find the design here: https://zpl.io/bL6zB7J
any questions let me know! thanks

**Comment by ClementChaumel** [comment]
> @ClementChaumel find the design here: https://zpl.io/bL6zB7J
> any questions let me know! thanks

Does this button still open the collapsible table or is it a direct link to /subscribe?
image

**Comment by wgx** [comment]
@anasereijo @ClementChaumel Just link out to /subscribe (the subscription builder / checkout)

**Comment by ClementChaumel** [comment]
This has changed quite a bit and now include a toggle for auto renewal.
It's not doing anything yet but I've left a couple of TODOs indicating what needs to be done. (Basically add some logic to the endpoint created to send data to the contracts backend)

**Comment by wgx** [comment]
Demo'd with @ClementChaumel just now - +1

File: static/js/src/advantage/auto-renewal.js [hunk 6]
```diff
@@ -0,0 +1,45 @@
...
+    autoRenewalSection.setAttribute("aria-hidden", "false");
+  }
+}
+
+function confirmChanges() {
+  const enabledRadio = document.getElementById("auto-renewal-on");
+  const subscriptionId = this.dataset.subscriptionId;
+
+  setAutoRenewal(subscriptionId, enabledRadio.checked).catch((error) => {
+    console.error(error);
```
Comment (anthonydillon) [inline static/js/src/advantage/auto-renewal.js:25]: Can this push an error to raven so we see it in sentry.

File: static/js/src/advantage/contracts-api.js [hunk 7]
```diff
@@ -222,3 +222,24 @@ export async function postCustomerInfoForPurchasePreview(
   let data = await response.json();
   return data;
 }
+
+export async function setAutoRenewal(subscriptionId, value) {
+  const queryString = window.location.search;
```
Comment (anthonydillon) [inline static/js/src/advantage/contracts-api.js:227]: I assume this is just for `test_backend` maybe a comment here would make to clearer in the future

File: static/sass/styles.scss [hunk 8]
```diff
@@ -535,6 +535,15 @@ summary {
       top: -2rem;
     }
   }
+
+  &--center {
```
Comment (anthonydillon) [inline static/sass/styles.scss:539]: I don't think you need this anymore

File: templates/advantage/index.html [hunk 9]
```diff
@@ -40,22 +40,44 @@ <h1>Ubuntu Advantage for&nbsp;Infrastructure</h1>
...
+        <div class="row">
+          <div class="col-8 col-medium-3">
+            <p class="u-no-margin--bottom">Your next recurring payment of <strong><span class="js-format-price">{{subscriptions["next_payment"]["ammount"]}}</span></strong> will be taken on <strong>{{subscriptions["next_payment"]["date"]}}</strong>.</p>
+          </div>
+          <div class="col-4 col-medium-3 u-align--right">
+            <div id="reveal-renewal-options">
+              <button class="p-button u-no-margin--bottom js-reveal-renewal-options" aria-controls="renewal-options" aria-expanded="false">Auto-renewal settings <i class="p-icon--contextual-menu">Open</i></button>
+            </div>
+          </div>
+          <div class="col-12  p-strip--light is-shallow u-hide p-card"  aria-hidden="true" id="renewal-options" style="border-left: 0; border-right: 0; overflow: visible; padding-top: 3rem; padding-bottom: 2rem;">
```
Comment (anthonydillon) [inline templates/advantage/index.html:56]: You can remove `col-12` here

**Review by anthonydillon — APPROVED** [review]
Few very simple comments

**Comment by anthonydillon** [comment]
Looks very good by the way!

File: static/js/src/advantage/contracts-api.js (hunk 7 above)
Comment (ClementChaumel) [reply static/js/src/advantage/contracts-api.js:227]: I added this as way to pass query params to the backend. It's currently only used for `test_backend` but could possibly be used for more. I'll add a comment

File: static/js/src/advantage/auto-renewal.js (hunk 6 above)
Comment (ClementChaumel) [reply static/js/src/advantage/auto-renewal.js:25]: Sentry logging seem to be done in the flask backend so I've added a skeleton there

File: webapp/views.py [hunk 10]
```diff
@@ -920,6 +984,25 @@ def post_anonymised_customer_info():
...
+    #     flask.current_app.extensions["sentry"].captureException(
+    #         extra={
+    #             "purchase_request": set_auto_renewal,
+    #             "api_response": http_error.response.json(),
+    #         }
+    #     )
+    #     return (
+    #         flask.jsonify({"error": "could change auto renewal settings"}),
+    #         500,
+    #     )
```
Comment (ClementChaumel) [inline webapp/views.py:1001]: Here is the sentry bit I mentioned @anthonydillon

File: static/sass/styles.scss (hunk 8 above)
Comment (ClementChaumel) [reply static/sass/styles.scss]: Good catch I've cleaned up the whole class

**Pull request merged**